packages/kotti-input-number/src/InputNumber.vue

```diff
@@ -137,70 +148,91 @@ export default {
 	},
 	watch: {
 		value: {
-			handler(newNumber) {
+			handler(newNumber, oldNumber) {
 				const newString = toString(newNumber)
-				const shouldUpdate = this.currentValueNumber !== toNumber(newString)
+				const truncatedNumber = toNumber(newString)
+				const { min, max, step } = this
+
+				if (!isInRange({ max, min, offset: 0, value: truncatedNumber }))
+					throw new RangeError(
+						`KtInputNumber: encounted an out-of-range number "${newNumber}"`,
+					)
 
-				if (shouldUpdate) this.setValue(newString)
+				if (!isStepMultiple({ min, step, value: truncatedNumber }))
+					throw new Error(
+						`KtInputNumber: encounted a value "${newNumber}" that doesn't fit ((min + k * step): where k is an integer)`,
+					)
+
+				const shouldUpdate = oldNumber !== truncatedNumber
+				if (shouldUpdate) this.setInternalStringValue(newString)
 			},
 			immediate: true,
 		},
 	},
 	methods: {
 		decrementValue() {
 			if (!this.isDecrementEnabled) return
-			if (this.currentValueNumber === null) {
-				const defaultString = this.min === null ? '0' : this.min + ''
-				this.setValue(defaultString)
+			if (this.value === null) {
+				const defaultNumber = this.min || 0
+				this.emitInput(defaultNumber)
 				return
 			}
-
-			this.setValue(toString(this.currentValueNumber - this.step))
+			this.emitInput(this.value - this.step)
+		},
+		emitInput(value) {
+			this.$emit('input', value)
 		},
 		handleBlur() {
-			this.setValue(toString(this.currentValueNumber))
+			this.setInternalStringValue(toString(this.value))
 		},
 		handleInput(value) {
-			const { max, min } = this
+			const { max, min, step } = this
 
 			const valueWithoutLeadingZeroes = value.replace(
 				LEADING_ZEROES_REGEX,
 				'$1',
 			)
+			const nextNumber = toNumber(valueWithoutLeadingZeroes)
 
 			const isTypedNumberValid =
 				VALID_REGEX.test(valueWithoutLeadingZeroes) &&
-				(STRINGS_THAT_ARE_TREATED_AS_NULL.includes(valueWithoutLeadingZeroes) ||
-					isInRange({ max, min, value: toNumber(valueWithoutLeadingZeroes) }))
-
-			if (isTypedNumberValid) this.setValue(valueWithoutLeadingZeroes)
-			else this.hasFormError = true
-
+				isStepMultiple({
+					min,
+					step,
+					value: nextNumber,
+				}) &&
+				isInRange({
+					max,
+					min,
+					offset: 0,
+					value: nextNumber,
+				})
+
+			if (isTypedNumberValid) {
+				const shouldEmit = nextNumber !== this.value
+				if (shouldEmit) this.emitInput(nextNumber)
+				else this.setInternalStringValue(valueWithoutLeadingZeroes)
```

Comment:
`toNumber` has a `parseFloat` call which would treat `12.` as `12` for example, so if we would emit, we would prevent the user from ever typing `"12."` , because we will emit the parsed `12` instead which will trigger the watcher, then we set the `internalStringValue` to `"12"` again. 
nevertheless, we want to set the `internalStringValue` to bind it to the `value` of the `<input>` (check template), to show the `12.`